src/app/InteractionModelEngine.cpp

```diff
@@ -462,40 +397,96 @@ void InteractionModelEngine::OnResponseTimeout(Messaging::ExchangeContext * ec)
                     ChipLogValueExchange(ec));
 }
 
-CHIP_ERROR InteractionModelEngine::SendReadRequest(ReadPrepareParams & aReadPrepareParams, ReadClient::Callback * aCallback)
+uint16_t InteractionModelEngine::GetWriteClientArrayIndex(const WriteClient * const apWriteClient) const
 {
-    ReadClient * client = nullptr;
-    CHIP_ERROR err      = CHIP_NO_ERROR;
-    ReturnErrorOnFailure(NewReadClient(&client, ReadClient::InteractionType::Read, aCallback));
-    err = client->SendRequest(aReadPrepareParams);
-    if (err != CHIP_NO_ERROR)
-    {
-        client->Shutdown();
-    }
-    return err;
+    return static_cast<uint16_t>(apWriteClient - mWriteClients);
 }
 
-CHIP_ERROR InteractionModelEngine::SendSubscribeRequest(ReadPrepareParams & aReadPrepareParams, ReadClient::Callback * aCallback)
+uint16_t InteractionModelEngine::GetReadHandlerArrayIndex(const ReadHandler * const apReadHandler) const
 {
-    ReadClient * client = nullptr;
-    ReturnErrorOnFailure(NewReadClient(&client, ReadClient::InteractionType::Subscribe, aCallback));
-    ReturnErrorOnFailure(client->SendRequest(aReadPrepareParams));
-    return CHIP_NO_ERROR;
+    return static_cast<uint16_t>(apReadHandler - mReadHandlers);
 }
 
-uint16_t InteractionModelEngine::GetReadClientArrayIndex(const ReadClient * const apReadClient) const
+void InteractionModelEngine::AddReadClient(ReadClient * apReadClient)
 {
-    return static_cast<uint16_t>(apReadClient - mReadClients);
+    ReadClient * pPrevListItem = nullptr;
+    ReadClient * pCurListItem  = mpActiveReadClientList;
+
+    while (pCurListItem)
+    {
+        pPrevListItem = pCurListItem;
+        pCurListItem  = pCurListItem->GetNextClient();
+    }
+
+    if (pPrevListItem)
+    {
+        pPrevListItem->SetNextClient(apReadClient);
+    }
+    else
+    {
+        mpActiveReadClientList = apReadClient;
+    }
+
+    apReadClient->SetNextClient(nullptr);
```

Comment:
Does the order here matter?  Because it seems it might be simpler to do:
```suggestion
    apReadClient->SetNextClient(mpActiveReadClientList);
    mpActiveReadClientList = apReadClient;
```